### ISSUE TITLE
`npm install` always fails with "gclient_runhooks failed: 2 null"

### ISSUE THREAD

**Issue by bawNg** [node-webrtc/node-webrtc#45]
I have tried installing the `wrtc` package as well as from the develop branch source unsuccessfully. Building always fails with the following error:

```
~/git/node-webrtc$ npm install

> wrtc@0.0.6 install /home/admin/git/node-webrtc
> node build.js

gclient_config
.
gclient_config complete
gclient_sync
....
gclient_sync complete
gclient_runhooks
........
gclient_runhooks failed: 2 null
```

**Comment by modeswitch** [comment]
Could you try running `gclient runhooks` in `lib/libwebrtc` and see what output you get? Also, which platform are you on?

**Comment by bawNg** [comment]
I needed to install libgtk2.0-dev and libnss3-dev to get `gclient runhooks` working on Debian Squeeze 6.0.6.

Now `npm install` fails at "ninja_build failed: 1 null". I saw you told someone else in another issue to use ninja manually so I tried that to get more info on the failure.

```
$ ninja -C lib/libwebrtc/trunk/out/Release
ninja: Entering directory `lib/libwebrtc/trunk/out/Release'
[18/2357] CC obj/third_party/openssl/openssl/crypto/poly1305/openssl.poly1305_vec.o
FAILED: gcc -MMD -MF obj/third_party/openssl/openssl/crypto/poly1305/openssl.poly1305_vec.o.d -DANGLE_DX11 -D_FILE_OFFSET_BITS=64 -DV8_DEPRECATION_WARNINGS -DCHROMIUM_BUILD -DUSE_CAIRO=1 -DUSE_GLIB=1 -DUSE_DEFAULT_RENDER_THEME=1 -DUSE_LIBJPEG_TURBO=1 -DUSE_NSS=1 -DUSE_X11=1 -DENABLE_ONE_CLICK_SIGNIN -DGTK_DISABLE_SINGLE_INCLUDES=1 -DUSE_XI2_MT=2 -DENABLE_REMOTING=1 -DENABLE_WEBRTC=1 -DENABLE_PEPPER_CDMS -DENABLE_CONFIGURATION_POLICY -DENABLE_INPUT_SPEECH -DENABLE_NOTIFICATIONS -DUSE_UDEV -DENABLE_EGLIMAGE=1 -DENABLE_TASK_MANAGER=1 -DENABLE_EXTENSIONS=1 -DENABLE_PLUGIN_INSTALLATION=1 -DENABLE_PLUGINS=1 -DENABLE_SESSION_SERVICE=1 -DENABLE_THEMES=1 -DENABLE_BACKGROUND=1 -DENABLE_AUTOMATION=1 -DENABLE_GOOGLE_NOW=1 -DCLD_VERSION=2 -DENABLE_FULL_PRINTING=1 -DENABLE_PRINTING=1 -DENABLE_SPELLCHECK=1 -DENABLE_CAPTIVE_PORTAL_DETECTION=1 -DENABLE_MANAGED_USERS=1 -DENABLE_MDNS=1 -DNO_WINDOWS_BRAINDEATH -DPURIFY -DMONOLITH '-DENGINESDIR="/dev/null"' '-DOPENSSLDIR="/etc/ssl"' -DOPENSSL_BN_ASM_GF2m -DOPENSSL_BN_ASM_MONT -DAES_ASM -DGHASH_ASM -DSHA1_ASM -DSHA256_ASM -DSHA512_ASM -DMD5_ASM -DDES_PTR -DDES_RISC1 -DDES_UNROLL -DOPENSSL_CPUID_OBJ -DNDEBUG -DNVALGRIND -DDYNAMIC_ANNOTATIONS_ENABLED=0 -I../../third_party/openssl/config/x64 -I../../third_party/openssl -I../../third_party/openssl/openssl -I../../third_party/openssl/openssl/crypto -I../../third_party/openssl/openssl/crypto/asn1 -I../../third_party/openssl/openssl/crypto/evp -I../../third_party/openssl/openssl/crypto/modes -I../../third_party/openssl/openssl/include -fstack-protector --param=ssp-buffer-size=4 -pthread -fno-exceptions -fno-strict-aliasing -Wno-unused-parameter -Wno-missing-field-initializers -fvisibility=hidden -pipe -fPIC -w -Wno-format -Wno-unused-result -O2 -fno-ident -fdata-sections -ffunction-sections   -c ../../third_party/openssl/openssl/crypto/poly1305/poly1305_vec.c -o obj/third_party/openssl/openssl/crypto/poly1305/openssl.poly1305_vec.o
../../third_party/openssl/openssl/crypto/poly1305/poly1305_vec.c:72: error: expected â€˜=â€™, â€˜,â€™, â€˜;â€™, â€˜asmâ€™ or â€˜__attribute__â€™ before â€˜uint128_tâ€™
../../third_party/openssl/openssl/crypto/poly1305/poly1305_vec.c:80: error: expected â€˜=â€™, â€˜,â€™, â€˜;â€™, â€˜asmâ€™ or â€˜__attribute__â€™ before â€˜inlineâ€™
../../third_party/openssl/openssl/crypto/poly1305/poly1305_vec.c:86: error: expected â€˜=â€™, â€˜,â€™, â€˜;â€™, â€˜asmâ€™ or â€˜__attribute__â€™ before â€˜inlineâ€™
../../third_party/openssl/openssl/crypto/poly1305/poly1305_vec.c:92: error: expected â€˜=â€™, â€˜,â€™, â€˜;â€™, â€˜asmâ€™ or â€˜__attribute__â€™ before â€˜inlineâ€™
../../third_party/openssl/openssl/crypto/poly1305/poly1305_vec.c:99: error: expected â€˜)â€™ before â€˜aâ€™
../../third_party/openssl/openssl/crypto/poly1305/poly1305_vec.c:105: error: expected â€˜)â€™ before â€˜vâ€™
../../third_party/openssl/openssl/crypto/poly1305/poly1305_vec.c: In function â€˜shr128_pairâ€™:
../../third_party/openssl/openssl/crypto/poly1305/poly1305_vec.c:113: error: â€˜uint128_tâ€™ undeclared (first use in this function)
../../third_party/openssl/openssl/crypto/poly1305/poly1305_vec.c:113: error: (Each undeclared identifier is reported only once
../../third_party/openssl/openssl/crypto/poly1305/poly1305_vec.c:113: error: for each function it appears in.)
../../third_party/openssl/openssl/crypto/poly1305/poly1305_vec.c:113: error: expected â€˜)â€™ before â€˜hiâ€™
../../third_party/openssl/openssl/crypto/poly1305/poly1305_vec.c: In function â€˜poly1305_first_blockâ€™:
../../third_party/openssl/openssl/crypto/poly1305/poly1305_vec.c:253: error: â€˜uint128_tâ€™ undeclared (first use in this function)
../../third_party/openssl/openssl/crypto/poly1305/poly1305_vec.c:253: error: expected â€˜;â€™ before â€˜dâ€™
../../third_party/openssl/openssl/crypto/poly1305/poly1305_vec.c:277: error: â€˜dâ€™ undeclared (first use in this function)
../../third_party/openssl/openssl/crypto/poly1305/poly1305_vec.c: In function â€˜CRYPTO_poly1305_finishâ€™:
../../third_party/openssl/openssl/crypto/poly1305/poly1305_vec.c:640: error: â€˜uint128_tâ€™ undeclared (first use in this function)
../../third_party/openssl/openssl/crypto/poly1305/poly1305_vec.c:640: error: expected â€˜;â€™ before â€˜dâ€™
../../third_party/openssl/openssl/crypto/poly1305/poly1305_vec.c:678: error: â€˜dâ€™ undeclared (first use in this function)
[18/2357] CC obj/third_party/icu/source/common/icuuc.uresbund.o
ninja: build stopped: subcommand failed.
```

**Comment by modeswitch** [comment]
Can you try removing everything in the `libjingle` directory and running

```
gclient sync
gclient runhooks
ninja -C trunk/out/Release peerconnection_client
```

Something may have gotten wedged from earlier, so this will start a clean build.

**Comment by bawNg** [comment]
It got stuck on a few other files before I removed everything in the `out` directory and after running those three commands again it now gets to `[908/2028] CC obj/third_party/openssl/openssl/ssl/openssl.t1_lib.o` before failing with `ninja: build stopped: subcommand failed.`.

**Comment by modeswitch** [comment]
I just ran through a clean build on my machine and it compiles and links `peerconnection_client`, no failures. I haven't tried building this on a debian box, but you may be missing some library prerequisites that for some reason aren't fetched by gclient.

The WebRTC implementation is maintained as a different project, so you may find some answers here: [here](http://www.webrtc.org/reference/getting-started) or [here](http://www.webrtc.org/reference/getting-started/prerequisite-sw). Give those a try and if you're still unable to make progress, please update this issue again and I'll see what else I can do to help.

**Comment by bawNg** [comment]
I started over on an Ubuntu 12.04 box and after sorting out the required dependencies I managed to get it building successfully. One issue I did notice with my initial testing is that I have seemingly random segmentation faults, usually after a few connection attempts.

Another issue is that the candidate ports that are chosen and passed to the `onicecandidate` event callback are random and there does not appear to be any way to force specific ports to be used. A peer running on a server requires static ports to be used as opening a huge number of ports on the firewall is not a viable option. This means that browser based peers who do not have any working NAT traversal options are unable to establish a connection to my server-side peer.

**Comment by modeswitch** [comment]
Please open new issues for those problems.

For the segfault, please provide the test program that you're using, and if you're able to get a backtrace that would be great also.

For the `onicecandidate` issue, NAT-traversal is required for using WebRTC (uPnP, or something). Those ports are returned by the underlying WebRTC implementation.

**Comment by piranna** [comment]
I have just downloaded the code from the repo and done a clean install and installation process crash on gclient_sync

``` bash
[piranna@Latitude:~/Dropbox/Proyectos/node-webrtc]
 > npm install

> wrtc@0.0.6 install /home/piranna/Dropbox/Proyectos/node-webrtc
> node build.js

gclient_config
.
gclient_config complete
gclient_sync
..
gclient_sync failed: 1 null
```

**Comment by piranna** [comment]
Ok, doesn't mind, it was because I didn't have subversion installed. This and depot_tools should be added as dependencies, so they get installed automatically, or checked and notified to the user previously to do the install.

Update:

Now 'gclient runhooks' is also failling to me, saying it can't be able to found "include/jni.h" file. I've installed openjdk 6 & 7 without results :-(

**Comment by piranna** [comment]
After some tests, I've done a clean up and on LinuxMint 16 (Ubuntu 13.10) I've done the next steps and got an error of `nodegyp_build failed: 1 null`:

``` bash
# deptot tools
git clone https://chromium.googlesource.com/chromium/tools/depot_tools.git
export PATH="$PATH":`pwd`/depot_tools

# Developer libraries
sudo apt-get install libgtk2.0-dev libnss3-dev libasound2-dev libpulse-dev libssl-dev

# Java
sudo apt-get install openjdk-7-jdk
export JAVA_HOME=/usr/lib/jvm/default-java

# Do the install
git clone git@github.com:ShareIt-project/node-webrtc.git
npm install
```

**Comment by piranna** [comment]
Ok, I've found what's the error, seems that's trying to create an instance of the PeerConnection class while it's abstract. In any case, should PeerConnection be RTCPeerConnection? They renamed it a long time ago...

Launching `node-gyp build` manually give this:

``` bash
[piranna@Latitude:~/Dropbox/Proyectos/node-webrtc]
 > node_modules/.bin/node-gyp build
gyp info it worked if it ends with ok
gyp info using node-gyp@0.10.10
gyp info using node@0.10.24 | linux | x64
gyp info spawn make
gyp info spawn args [ 'BUILDTYPE=Release', '-C', 'build' ]
make: se ingresa al directorio «/home/piranna/Dropbox/Proyectos/node-webrtc/build»
  CXX(target) Release/obj.target/webrtc/src/peerconnection.o
../src/peerconnection.cc: In static member function ‘static v8::Handle<v8::Value> PeerConnection::New(const v8::Arguments&)’:
../src/peerconnection.cc:351:44: error: cannot allocate an object of abstract type ‘PeerConnection’
   PeerConnection* obj = new PeerConnection();
                                            ^
In file included from ../src/peerconnection.cc:17:0:
../src/peerconnection.h:77:7: note:   because the following virtual functions are pure within ‘PeerConnection’:
 class PeerConnection
       ^
In file included from ../src/peerconnection.cc:14:0:
../lib/libwebrtc/trunk/talk/app/webrtc/peerconnectioninterface.h:280:16: note:  virtual void webrtc::PeerConnectionObserver::OnRenegotiationNeeded()
   virtual void OnRenegotiationNeeded() = 0;
                ^
make: *** [Release/obj.target/webrtc/src/peerconnection.o] Error 1
make: se sale del directorio «/home/piranna/Dropbox/Proyectos/node-webrtc/build»
gyp ERR! build error 
gyp ERR! stack Error: `make` failed with exit code: 2
gyp ERR! stack     at ChildProcess.onExit (/home/piranna/Dropbox/Proyectos/node-webrtc/node_modules/node-gyp/lib/build.js:267:23)
gyp ERR! stack     at ChildProcess.EventEmitter.emit (events.js:98:17)
gyp ERR! stack     at Process.ChildProcess._handle.onexit (child_process.js:789:12)
gyp ERR! System Linux 3.11.0-12-generic
gyp ERR! command "node" "/home/piranna/Dropbox/Proyectos/node-webrtc/node_modules/.bin/node-gyp" "build"
gyp ERR! cwd /home/piranna/Dropbox/Proyectos/node-webrtc
gyp ERR! node -v v0.10.24
gyp ERR! node-gyp -v v0.10.10
gyp ERR! not ok 
```

**Comment by modeswitch** [comment]
Looks like a code change in the library. It's fixed in my local branch and I'll push it up to github soon.

**Comment by modeswitch** [comment]
Also, worth mentioning that things like this will happen from time to time while node-webrtc tracks the development branch of libjingle. Once all the features we need (<cough>sctp data channels</cough>) are in the stable branch we can switch to that and this will happen less often.

**Comment by piranna** [comment]
Please keep us informed about when we can download that fix, there are two issues regarding to browserify compatibility and clean-up of file hierarchy that I would like publish.

Would it make sense to have the library compiled somewhere in an independent project (libwebrtc) so node-webrtc is just a wrapper around it, instead of needed to download and compile the full library source code each time?

**Comment by modeswitch** [comment]
Fix is on github now.

#29 is intended to address the issue of having to build the library on common platforms.

**Comment by piranna** [comment]
Ok, now I have been able to make it work, thanks for the effort :-)

2014/1/17 Alan K notifications@github.com

> Fix is on github now.
> 
> #29 https://github.com/js-platform/node-webrtc/issues/29 is intended to
> address the issue of having to build the library on common platforms.
> 
> —
> Reply to this email directly or view it on GitHubhttps://github.com/js-platform/node-webrtc/issues/45#issuecomment-32571219
> .

## 

"Si quieres viajar alrededor del mundo y ser invitado a hablar en un monton
de sitios diferentes, simplemente escribe un sistema operativo Unix."
– Linus Tordvals, creador del sistema operativo Linux

**Comment by piranna** [comment]
Seems libwebrtc has changed and now doesn't compile again. We should consider to use a stable revision and upload it each week instead work always agains latest one.

**Comment by JessieAMorris** [comment]
By the way, for anyone hitting this issue in the future, what fixed it for me was installing gclient and other such depot_tools to my $PATH.

**Comment by piranna** [comment]
With the updated install script it should download it automatically.

2014/1/21 Jessie A. Morris notifications@github.com

> By the way, for anyone hitting this issue in the future, what fixed it for
> me was installing gclient and other such depot_tools to my $PATH.
> 
> —
> Reply to this email directly or view it on GitHubhttps://github.com/js-platform/node-webrtc/issues/45#issuecomment-32807119
> .

## 

"Si quieres viajar alrededor del mundo y ser invitado a hablar en un monton
de sitios diferentes, simplemente escribe un sistema operativo Unix."
– Linus Tordvals, creador del sistema operativo Linux

**Comment by JessieAMorris** [comment]
It downloaded it correctly (and placed it in tools/depot_tools) but it did not seem to place it on my $PATH for me. The runhooks would die with an error two.

**Comment by piranna** [comment]
Yeah, if you launch it by hand, it tells you that can't be able to find a
script to exec.

2014/1/21 Jessie A. Morris notifications@github.com

> It downloaded it correctly (and placed it in tools/depot_tools) but it did
> not seem to place it on my $PATH for me. The runhooks would die with an
> error two.
> 
> —
> Reply to this email directly or view it on GitHubhttps://github.com/js-platform/node-webrtc/issues/45#issuecomment-32807911
> .

## 

"Si quieres viajar alrededor del mundo y ser invitado a hablar en un monton
de sitios diferentes, simplemente escribe un sistema operativo Unix."
– Linus Tordvals, creador del sistema operativo Linux

**Comment by modeswitch** [comment]
@piranna This issue has is starting to cover too many things. Could you close it and open new issues for things that are still relevant?

**Comment by modeswitch** [comment]
@piranna I just synced and rebuilt without any errors. Could you retry and if you're still having problems open a new issue with the build output and I'll have a look.

**Comment by piranna** [comment]
I can't be able to close it, the close button has dissapeared :-/

Anyway, I've exec by hand all the commands done by the script and I'm getting the next trace on `gclient sync` after it has finished to download all the files:

```
________ running '/usr/bin/python trunk/build/cp.py trunk/../chromium_gn/.gn trunk' in '/home/piranna/Dropbox/Proyectos/node-webrtc/third_party/libwebrtc'

________ running 'download_from_google_storage --no_resume --platform=win32 --no_auth --bucket chromium-gn -s trunk/tools/gn/bin/win/gn.exe.sha1' in '/home/piranna/Dropbox/Proyectos/node-webrtc/third_party/libwebrtc'
Traceback (most recent call last):
  File "../../tools/depot_tools/gclient.py", line 1923, in <module>
    sys.exit(Main(sys.argv[1:]))
  File "../../tools/depot_tools/gclient.py", line 1913, in Main
    return dispatcher.execute(OptionParser(), argv)
  File "/home/piranna/Dropbox/Proyectos/node-webrtc/tools/depot_tools/subcommand.py", line 245, in execute
    return command(parser, args[1:])
  File "../../tools/depot_tools/gclient.py", line 1697, in CMDsync
    ret = client.RunOnDeps('update', args)
  File "../../tools/depot_tools/gclient.py", line 1244, in RunOnDeps
    self.RunHooksRecursively(self._options)
  File "../../tools/depot_tools/gclient.py", line 822, in RunHooksRecursively
    hook, cwd=self.root.root_dir, always=True)
  File "/home/piranna/Dropbox/Proyectos/node-webrtc/tools/depot_tools/gclient_utils.py", line 244, in CheckCallAndFilterAndHeader
    return CheckCallAndFilter(args, **kwargs)
  File "/home/piranna/Dropbox/Proyectos/node-webrtc/tools/depot_tools/gclient_utils.py", line 438, in CheckCallAndFilter
    **kwargs)
  File "/home/piranna/Dropbox/Proyectos/node-webrtc/tools/depot_tools/subprocess2.py", line 237, in __init__
    super(Popen, self).__init__(args, **kwargs)
  File "/usr/lib/python2.7/subprocess.py", line 709, in __init__
    errread, errwrite)
  File "/usr/lib/python2.7/subprocess.py", line 1326, in _execute_child
    raise child_exception
OSError: [Errno 2] No existe el archivo o el directorio
```

**Comment by modeswitch** [comment]
@piranna I added you to the project, so you should be able to manage issues now. I'm definitely not seeing that gclient issue though.

**Comment by modeswitch** [comment]
Closing. Please track build failures in #72 for now.
